Feature/support env hcl and interpolations

XT-Phate opened the pull request:
Hello,

This is a proposition to support HCL files as values inputs with interpolation.
I didn't want to come empty handed while making this proposition, so here is a PR.


## Context : 
I love helmfile in its concept but there was a missing part on how I'd like to use it and as I think many people may want to use it.  
When working with a complex and heavy project that requires many helm charts, there can be many dependencies between them or even logic that'd have an impact on 2 different apps. I assume that's why the `need` option was created
But that's also true in terms of variables.  
Let's assume the following case using the current global env values :   

db_values.yaml.gotmpl  
```yaml
db_name: postgres
db_namespace: db 
#Option 1
postgres_kubernetes_url: "{{ .Values.db_name }}.{{ .Values.db_namespace }}"
# Option 2
{{ $conn := printf "%s%s" .Values.db_name  .Values.db_namespace  }} #Only live in the local template
postgres_kubernetes_url: {{ $conn }}
```

app1.yaml.gotmpl  
```yaml
app1_a: a
app1_b: b
app1_db_url: "{{ .Values.postgres_kubernetes_url }}"
```

With Option 1, you'll need to go through different `---` seperator for each value file  
With Option 2, you'll need to repeat your vars multiple times and the content of $ vars can't be shared across files.  
And ofc, this is getting trickier as the number of releases grows. Especially if apps becomes more dependant on each other like that :  

app2.yaml.gotmpl  
```yaml
app2_a: "{{ .Values.global_a }}{{ .Values.app1_a }}{{ .Values.app1_b }}"
```


While this is doable by layering the Values files, this can get very tricky when building a lot of new vars dependent from each other in different files and this create a hard to maintain helmfile.

So that's why I've come up with integrating the HCL syntax and a sorted interpolation system based on the already implemented DAG used for the release order.
	

## Current specs (updated)

* helmfile hcl  files should always be identified with the `.hcl` suffix
* hcl variables should follow the hcl syntax
* helmfile hcl values will only be parsed when set into a `values` block or in a `locals` block
* `values` block is a shared value block where all vars are accessible everywhere in all loaded files
* `locals` is its own block, which can't reference external variables apart from the block itself, and where its defined vars are only accessible in its local file
* Only variables in `values` are made available to the root `.Values` (e.g : ` values { myvar = "var" }` is accessed through `{{ .Values.myvar }}`)
* There can only be 1 local block/file
* helmfile hcl `values` are referenced using the `hv` accessor.
* helmfile hcl `locals` are referenced using the `local` accessor.
* duplicated variables across .hcl `values` blocks are forbidden (An error will pop up specifying where are the duplicates)
* all hcl default functions are available and custom functions could be created in the future
* hemlfile hcl files can be encrypted. The same process as environments values applies
* HCL variables have a higher precedence over the .yaml or .gotmpl files (this is indeed debatable and could be "optionized" to have the reversed option). But are superseeded by any yaml secret files. `clear Yaml values < hcl == hcl encrypted < secrets yaml values `


## Input Example (more complete one in the provided test function) 
helmfile.yaml
```yaml
environments:
  default:
    values:
      - values.hcl
```
values.hcl
```terraform
locals {
  myVar = "var"
}
values {
  varA = "A"
  varB = "${hv.A}-B"
  varC = local.myVar
}
```

Happy to get any feedback/suggestion/correction ! And if you feel like this has nothing to do in helmfile, don't hesitate also to tell me.

mumoshu commented on pkg/state/testdata/values.7.hcl:
```diff
@@ -0,0 +1,39 @@
+helmfile_vars {
```
FWIW, I'm a bit concerned about we now have no way to define local variables like `{{ $foo := ... }}".

Similar to private vs public variables in a standard programming language, I often used the `$name` vars as local variables that are not needed to be reused across files.

I believe your motivation for this HCL feature is to effortlessly share and reference some variables across files. The use of HCL for deferred parsing combined with our DAG code solves this. Awesome!

But private vs public distinction is just missing in the picture maybe?

Can we perhaps introduce something like terraform's "locals" and let users use it as a set of local variables that may or may not be exposed as helm values, and use "helmfile_vars" (or "values" as I proposed separately) to expose them?

```hcl
// locals can be shared across hcl files
// however it is not exposed as helm values
locals {
  a = "a"
  b = "b"
  c = "${local.a}${local.b}"
}

// you expose helm values with values blocks
values {
  // expose a local as-is
  a = local.a
  // use a local to render a helm value
  prefiexed_c = "someprefix${local.c}"
}

// values are {"a":"a","prefiexed_c ":"someprefixab"}

// we can even have a way to append the entire hcl object to helm values :)
locals {
  temp_values = tomap({
    c= local.c
    d= "d"
  })
}

append {
  values = locals.temp_values
}

// values are now {"a":"a","prefiexed_c ":"someprefixab", "c":"ab","d":"d"}
```

XT-Phate replied:
Indeed, I didn't have this private vs public distinction in mind as I was mainly inspired by terraform for this which doesn't really have it either.

```
Similar to private vs public variables in a standard programming language, I often used the $name vars as local variables that are not needed to be reused across files.
```
From what I understand, your goal is to have a local "name" variable for example, and so that it only lives inside the file so that it doesn't have any conflict elsewhere and avoid the var naming complexity ?
	
If we were to do that, I'd definitely keep the variables accessors for all vars which would go against your previous remark (otherwise if someone would set a value named "local", that'd break it)
Please note that the values, from the go templates, are accessed through the .Values.<var> without the specified accessor which only exist in the hcl values world

So this is how I would translate it: 

* locals block would all be evaluated first before anything as they are indenpendent block and can't have any external dependencies

* In practice : 
file1.hcl
```terraform
values {
  a = "a"
  b = "b"
}
```
file2.hcl
```terraform
// I'd need to check how far I could push it, but I'm not sure having multiple locals blocks would be easy to handle in the same file based on your spec
locals {
  c = "a"
  // This wouldn't work as locals cannot be part of the global gaph in this context
  d = "${hv.a} - ${hv.b}" 
  e = local.c // OK
}

values {
  // As we are aware at runtime in which file the variable is through the hcl.Range object
  // That also means that all locals block are evaluated before any `values` block
  // we can inject the corresponding local block reference in a map[string]map[string]*Values (first map indexes the filename, 2nd the variable)
  val = "${local.c} - ${hv.a}"
}


// Regarding that, if you'd want to expose the locals like this, 
// I don't think of a proper way than to put them in their own vars because appending it raw like this could overwrite existing vars

append {
  values = locals.temp_values
}

// Which'd rather look like this somehow
values {
	myLocals = local
}

```

XT-Phate replied:
Alright, so I just pushed it this way. I'm updating my original post accordingly regarding the specs

mumoshu reviewed (dismissed):
I did an initial review! Awesome work @XT-Phate!
I have experience using HCL as the config language in other tools and can see values in it! I hope we can land this in good shape.

mumoshu commented:
Oops, sorry, I didn't mean to approve it now! Take these as comments.

XT-Phate commented on pkg/state/hcl_loader.go:
```diff
@@ -0,0 +1,288 @@
+
+const (
+	badIdentifierDetail         = "A name must start with a letter or underscore and may contain only letters, digits, underscores, and dashes."
+	helmfileVarsBlockIdentifier = "helmfile_vars"
```
Sure ! no problem about it. 
I originally set "variables" because semantically to me, it "sounds" different than a "value". A value is static to me and a variable has a dynamic connotation. But "values" indeed does make more sense in helmfile context. I'll update it accordingly

yxxhero commented on pkg/state/hcl_loader.go:
```diff
@@ -0,0 +1,358 @@
+	//Generate Dag Plan which will provide the order from which to interpolate vars
+	plan, err := dagGraph.Plan(dag.SortOptions{
+		WithDependencies: true,
+	})
```
if err == nil {
   return &plan, nil
}

// err != nil

yxxhero commented on pkg/state/hcl_loader.go:
```diff
@@ -0,0 +1,358 @@
+	if root == localsAccessorPrefix && blockType != localsBlockIdentifier {
+		return "", nil
+	}
+	rootRange := traversal[0].SourceRange()
```
is ti safe? traversal[0]

XT-Phate replied:
In this specific context, yes, pretty much (the function is not exported)
A traversal will always at least have 1 _traverser_
Example : 
`myvar = local.var`
The _traversal_ "local.var" object for myvar contains 2 _traversers_ : `local` and `var`.
The HCL parser would fail before if a variable was set like this 
`myvar = `
if we imagine than it would be interpreted as a nil traverser, which it is not.

There is one way that'd make it not safe though that is crossing my mind now : if a future contributor would purposely pass an empty traversal to the function (even if it'd not make much sense tbh). I'll add something to avoid that. Thank you

yxxhero commented on pkg/state/hcl_loader.go:
add more unittest for funcions. thanks so much.

XT-Phate replied:
I added some that I think may be relevant. Tell me if you think it'd need more. Thanks

yxxhero commented:
@XT-Phate please fix conflic issue.

XT-Phate commented:
@yxxhero @mumoshu  I changed things a little bit :
* Move all the HCL related files to a `hcllang` package. That  should make it easier to review (commit [7105aeb](https://github.com/helmfile/helmfile/pull/1423/commits/7105aeb0ccdcd8647a45901cf466b97f994131f4) )  
* All the CTY standard go functions library. I thought of adding some from the OpenTofu project (like the encoding and files ones) as they are MPL2.0 licensed later. But this is already pretty heavy.
* The support of non-yaml secret files - I had to make sure not to impact the current precedence of the Environment `values` ( commit [7413341](https://github.com/helmfile/helmfile/pull/1423/commits/74133417543043839e4aae89cd03e6d8e4d0c13e) )

yxxhero commented:
@XT-Phate thanks so much. could you add an e2e test for this feature?

XT-Phate commented:
> @XT-Phate thanks so much. could you add an e2e test for this feature?
Sorry, It took me some time 
I just added it. 
I also added an integration one for the `secrets`. Which made me notice that it doesn't support helm secrets < 4.0.0. So I don't expect the CI to fully work. This is only valid for HCL secrets. The rest seems OK.
There is also that comment about helm secrets in the integration tests ci :
https://github.com/helmfile/helmfile/blob/main/.github/workflows/ci.yaml#L65-L69
```
            # We assume that the helm-secrets plugin is supposed to
            # work with the two most recent helm minor versions.
            # Once it turned out to be not practically true,
            # we will mark this combination as failable,
            # and instruct users to upgrade helm and helm-secrets at once.
```

yxxhero commented:
@XT-Phate please  update the doc for the feature.

yxxhero commented on docs/index.md:
```diff
@@ -967,55 +974,54 @@ releases:
-
-Prior to this [pull request](https://github.com/roboll/helmfile/pull/647), environment values were made available through the `{{ .Environment.Values.foo }}` syntax.
-This is still working but is **deprecated** and the new `{{ .Values.foo }}` syntax should be used instead.
+Since Helmfile vx.xxx.x, HCL language is supported for environment values only.
```
v0.164.0.

yxxhero commented:
@XT-Phate please fix confict and update doc. Then merge.

yxxhero commented:
ping

XT-Phate commented:
@yxxhero all done (y), thank you !
